### ISSUE TITLE
[java] Allow MissingOverride rule to work without bytecode

### ISSUE THREAD

**Issue by logesh14** [pmd/pmd#2428]
**Affects PMD Version:** 
6.20.0

**Rule:**
MissingOverride
Documentation url : https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#missingoverride

**Description:**
    MissingOverride rule is not working as described in the document.

**Code Sample demonstrating the issue:**

```
public class Foo implements Runnable {
                // This method is overridden, and should have an @Override annotation
                public void run() {

                }
            }
```

**Expected outcome:**
 As described in the document violation should be thrown for the method run() since it is overridden and also not annotated. 

It seems like it is false-negative. This an humble request to fix this issue.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

**Comment by pzygielo** [comment]
It works, but uses bytecode for analysis, so you have to use `-auxclasspath`, and then you get:
```shell
$ run.sh pmd -f text -rulesets category/java/bestpractices.xml/MissingOverride -d . -min 3 -auxclasspath .
Apr 21, 2020 3:08:19 PM net.sourceforge.pmd.PMD encourageToUseIncrementalAnalysis
WARNING: This analysis could be faster, please consider using Incremental Analysis: https://pmd.github.io/pmd-6.22.0/pmd_userdocs_incremental_analysis.html
Foo.java:4:	The method 'run()' is missing an @Override annotation.
```

**Comment by logesh14** [comment]
Thanks a lot @pzygielo. But when it comes to our codecheck standard if the java code contains this violation it shouldn't be allowed to push to the remote repository. Any chances that this rule can be analysed over source java without using the bytecodes in upcoming releases?

**Comment by jsotuyod** [comment]
@logesh14 no chance at all to do this properly without a complete classpath.

1. Your code does not only override methods defined by your own project, but also:
   1. for classes / interfaces defined by Java itself (as your own example)
   2. by third party dependencies you take in directly as compiled jars
2. PMD analyses one file at a time, and even with incrementalAnalysis it skips non-modified files, so it can not rely on looking at other files to know what types / methods they define

The fact that you need to check before push is independent on this. It's simply a matter of how you perform such check. Running PMD with the auxclasspath is just as feasible at any given point in your workflow.

You didn't fill your issue properly, so I don't know how you are running PMD, but if you are using any build tool (ant, maven, gradle) this is particularly easy. If you are using CLI, it will require a somewhat more cumbersome script, to compile java classes, create the auxclasspath file and then run PMD, but it's still doable.

**Comment by oowekyala** [comment]
In PMD 7 this will be partially supported. Since #2689 we only use bytecode to resolve external dependencies, but we don't need it for the current file. So in your example, `Foo.class` may be missing, but since `Runnable.class` will be available (it's from the JDK), the override will be correctly detected. If it wasn't implementing Runnable but some interface that is not on the classpath and not defined in the same file, there would be a false negative as today.

Anything more is challenging to implement given the way PMD currently works (file by file, like @jsotuyod explained). But it wouldn't just help MissingOverride, but all rules that use type resolution, so it would be nice to have that in the future. Since this is far bigger in scope than MissingOverride, I think we can close this.
